File: Platform.Converters.Tests/ConverterTests.cs
using System;
using Xunit;

namespace Platform.Converters.Tests
{
    public static class ConverterTests
    {
        [Fact]
        public static void SameTypeTest()
        {
            var result = UncheckedConverter<ulong, ulong>.Default.Convert(2UL);
            Assert.Equal(2UL, result);
            result = CheckedConverter<ulong, ulong>.Default.Convert(2UL);
            Assert.Equal(2UL, result);
        }

        [Fact]
        public static void Int32ToUInt64Test()
        {
            var result = UncheckedConverter<int, ulong>.Default.Convert(2);
            Assert.Equal(2UL, result);
            result = CheckedConverter<int, ulong>.Default.Convert(2);
            Assert.Equal(2UL, result);
        }

        [Fact]
        public static void SignExtensionTest()
        {
            var result = UncheckedSignExtendingConverter<byte, long>.Default.Convert(128);
            Assert.Equal(-128L, result);
            result = UncheckedConverter<byte, long>.Default.Convert(128);
            Assert.Equal(128L, result);
        }

        [Fact]
        public static void ObjectTest()
        {
            TestObjectConversion("1");
            TestObjectConversion(DateTime.UtcNow);
            TestObjectConversion(1.0F);
            TestObjectConversion(1.0D);
            TestObjectConversion(1.0M);
            TestObjectConversion(1UL);
            TestObjectConversion(1L);
            TestObjectConversion(1U);
            TestObjectConversion(1);
            TestObjectConversion((char)1);
            TestObjectConversion((ushort)1);
            TestObjectConversion((short)1);
            TestObjectConversion((byte)1);
            TestObjectConversion((sbyte)1);
            TestObjectConversion(true);
        }

        private static void TestObjectConversion<T>(T value) => Assert.Equal(value, UncheckedConverter<object, T>.Default.Convert(value));
    }
}

File: cpp/Platform.Converters.Tests/ConverterTests.cpp
namespace Platform::Converters::Tests
{
    TEST_CLASS(ConverterTests)
    {
        public: TEST_METHOD(SameTypeTest)
        {
            auto result = 2UL;
            Assert::AreEqual(2UL, result);
            result = CheckedConverter<std::uint64_t, std::uint64_t>.Default.Convert(2UL);
            Assert::AreEqual(2UL, result);
        }

        public: TEST_METHOD(Int32ToUInt64Test)
        {
            auto result = 2;
            Assert::AreEqual(2UL, result);
            result = CheckedConverter<std::int32_t, std::uint64_t>.Default.Convert(2);
            Assert::AreEqual(2UL, result);
        }

        public: TEST_METHOD(SignExtensionTest)
        {
            auto result = UncheckedSignExtendingConverter<std::uint8_t, std::int64_t>.Default.Convert(128);
            Assert::AreEqual(-128L, result);
            result = 128;
            Assert::AreEqual(128L, result);
        }

        public: TEST_METHOD(ObjectTest)
        {
            TestObjectConversion("1");
            TestObjectConversion(DateTime.UtcNow);
            TestObjectConversion(1.0F);
            TestObjectConversion(1.0D);
            TestObjectConversion(1.0M);
            TestObjectConversion(1UL);
            TestObjectConversion(1L);
            TestObjectConversion(1U);
            TestObjectConversion(1);
            TestObjectConversion((char)1);
            TestObjectConversion((std::uint16_t)1);
            TestObjectConversion((std::int16_t)1);
            TestObjectConversion((std::uint8_t)1);
            TestObjectConversion((std::int8_t)1);
            TestObjectConversion(true);
        }

        private: template <typename T> static void TestObjectConversion(T value) { Assert::AreEqual(value, value); }
    };
}
